### ISSUE TITLE
Restrict who can create batch changes (needs RBAC)

### ISSUE THREAD

**Issue by malomarrec** [sourcegraph/sourcegraph#34491]
### Problem statement

Some of our customers operate in regulated industries and have compliance requirements to **limit access to creating batch changes to only some users.**

### Problem validation / why 

Our initial reaction to this request was that :
- Permission issue: Batch Changes, when set up properly, map code hosts permissions and do not give more access to users than they already have without batch changes.
- PR spam issue: any developer can script out opening many changesets

We have verified that this was clear to those customers and confirmed that they still need to control access to batch changes to meet compliance goals. The customer's perception is that enabling developers to access a tool that makes it easier to automate creating many changesets increases risk and creates a compliance issue.

#### Other concerns

We were initially concerned that this could [slow down adoption](https://docs.google.com/document/d/1QkNLeLYOB_IwMp4s9Pd8kRiayI4EhvUp06XmDY9X3aQ/edit#heading=h.rvuii5ra4y31) and have a business impact. This was mitigated by making sure all users can _view_ batch changes, and only `creating` batch changes is optionally restricted.

### Customers

The customers are currently **blocked**:

- https://github.com/sourcegraph/accounts/issues/579
- https://github.com/sourcegraph/accounts/issues/1
- https://github.com/sourcegraph/accounts/issues/3
- https://github.com/sourcegraph/accounts/issues/8613

>There's no way we're going to give x,000 engineers the ability to deploy changes across all our repositories

- https://github.com/sourcegraph/accounts/issues/8243
Also see [product gap](https://sourcegraph2020.lightning.force.com/lightning/r/Product_Gap__c/a1B5b00000MwQNZEA3/view).

### Solution

**MVP:**

- Allow site-admins to optionally whitelist who can `create` batch changes. If a `restrictBatchChangesAccess` is set with a list of allowListed users, then creating Batch Changes is restricted to those users. What does that mean:
  - Running a preview with src-cli gives an error message `Creating batch changes has been restricted to some users only, you should contact your admin to get access`.
  - The create batch change button is disabled with an infobox and the same message
  - The Create batch change button on the search page is hidden (as well as any other buttons that prompt to create batch changes from another context)
  - All bulk actions / any edit action are disabled
  - The Batch Changes dashboard and batch change page are still visible, as well as non-edit `src batch` commands

**Comment by malomarrec** [comment]
Update: the IAM team is focusing on repo permissions this quarter, and not on RBAC building blocks. This is likely to be postponed until they get back to RBAC.

**Comment by malomarrec** [comment]
We've now reached a critical mass of requests on this issue. I think we can start designing it after 4.0, but implementation still depends on user management / RBAC being available or at least scoped out in order to avoid creating an inconsistent / unmaintainable admin experience.

**Comment by chrispine** [comment]
sounds like P2?

**Comment by malomarrec** [comment]
I'd say that the one thing we can work on this sprint, except the scope for 4.0 and current priorities.

**Comment by chrispine** [comment]
After scoping, we should create another ticket with the agreed-upon actual work

**Comment by malomarrec** [comment]
@chrispine should this be assigned to someone?

**Comment by chrispine** [comment]
@malomarrec I don't think anyone has capacity to do it this sprint, and it will get assigned to someone in the next sprint. If we want to move it into this sprint, we need to move something else out. But with vacation/conference/Lattice, we are already stretched pretty thin this sprint.

**Comment by chrispine** [comment]
assigned this to myself, as we need to decide what the priority of RBAC is and how that fits into strategic-readiness

**Comment by malomarrec** [comment]
@chrispine @BolajiOlajide what's the estimated timeline for this? The strategic-readiness board needs an estimated release version (added 4.2 for now, should it be 4.3)?

**Comment by BolajiOlajide** [comment]
This is dependent on #41940 - the rough estimate would be 4.3/4.4 given all the work that's to be done to get RBAC out of the way.

**Comment by courier-new** [comment]
Completed, pending docs!
